keycloak-event-emitter/src/main/java/io/cloudtrust/keycloak/kafkaeventemitter/KafkaEventEmitterProvider.java

```diff
@@ -0,0 +1,129 @@
+
+        produceEvent(buffer, adminEventTopic);
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private void completeEventAttributes(Event event) {
+        // add username if missing
+        if (event.getDetails() == null) {
+            event.setDetails(new HashMap<>());
+        }
+        String eventUsername = event.getDetails().get(Details.USERNAME);
+        if (!Strings.isNullOrEmpty(event.getUserId()) && Strings.isNullOrEmpty(eventUsername)) {
+            RealmModel realm = keycloakSession.realms().getRealm(event.getRealmId());
+            // retrieve username from userId
+            UserModel user = keycloakSession.users().getUserById(realm, event.getUserId());
+            if (user != null) {
+                event.getDetails().put(Details.USERNAME, user.getUsername());
+            }
+        }
+    }
+
+    private ExtendedAdminEvent completeAdminEventAttributes(IdentifiedAdminEvent adminEvent) {
+        ExtendedAdminEvent extendedAdminEvent = new ExtendedAdminEvent(adminEvent);
+        // add always missing agent username
+        ExtendedAuthDetails extendedAuthDetails = extendedAdminEvent.getAuthDetails();
+        if (!Strings.isNullOrEmpty(extendedAuthDetails.getUserId())) {
+            RealmModel realm = keycloakSession.realms().getRealm(extendedAuthDetails.getRealmId());
+            UserModel user = keycloakSession.users().getUserById(realm, extendedAuthDetails.getUserId());
+            extendedAuthDetails.setUsername(user.getUsername());
+        }
+        // add username if resource is a user
+        String resourcePath = extendedAdminEvent.getResourcePath();
+        if (resourcePath != null && resourcePath.startsWith("users")) {
+            // parse userID
+            String pattern = "^users/([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$";
+            Pattern r = Pattern.compile(pattern);
+            Matcher m = r.matcher(resourcePath);
+            if (m.matches()) {
+                String userId = m.group(1);
+                RealmModel realm = keycloakSession.realms().getRealm(adminEvent.getRealmId());
+                // retrieve user
+                UserModel user = keycloakSession.users().getUserById(realm, userId);
+                extendedAdminEvent.getDetails().put("user_id", userId);
+                if (user != null) {
+                    extendedAdminEvent.getDetails().put("username", user.getUsername());
+                }
+            }
+        }
+
+        return extendedAdminEvent;
+    }
+
+    private void produceEvent(ByteBuffer buffer, String topic) {
+        byte[] b = new byte[buffer.remaining()];
+        buffer.get(b);
+
+        String eventValue = Base64.getEncoder().encodeToString(b);
+
+        // Event production in Kafka topic
+        ProducerRecord<String, String> record = new ProducerRecord<>(topic, eventValue);
+        producer.send(record, new KafkaEventCallback(record));
```

Comment:
I think we can replace KafkaEventCallback with a lambda expression

Reply:
Done !